src/main.cpp

```diff
@@ -1,105 +1,144 @@
+        generator->generate_const_scalar_op("sqrt", type[0], type, "std::sqrt(a)");
+        generator->generate_const_scalar_op("exp", type[0], type, "std::exp(a)");
+        generator->generate_const_scalar_op("log", type[0], type, "std::log(a)");
         for (const char *prefix : {"", "a"}) {
             for (const char *suffix : {"", "h"}) {
                 for (const char *func : {"sin", "cos", "tan"}) {
                     std::ostringstream ss;
                     ss << prefix << func << suffix;
-                    generator.generate_const_scalar_op(
+                    generator->generate_const_scalar_op(
                         ss.str(), type[0], type, "std::" + ss.str() + "(a)");
                 }
             }
         }
     }
 
-    // Absolute value function.
-    generator.generate_const_scalar_op("abs", 'r', "r", "std::abs(a)");
-    generator.generate_const_scalar_op("abs", 'i', "i", "std::abs(a)");
-
-    // Complex number manipulation functions.
-    generator.generate_const_scalar_op("complex", 'c', "rr", "primitives::Complex(a, b)");
-    generator.generate_const_scalar_op("polar", 'c', "rr", "std::polar<double>(a, b)");
-    generator.generate_const_scalar_op("real", 'r', "c", "std::real<double>(a)");
-    generator.generate_const_scalar_op("imag", 'r', "c", "std::imag<double>(a)");
-    generator.generate_const_scalar_op("arg", 'r', "c", "std::arg<double>(a)");
-    generator.generate_const_scalar_op("norm", 'r', "c", "std::norm<double>(a)");
-    generator.generate_const_scalar_op("conj", 'c', "c", "std::conj<double>(a)");
+    // Absolute value function
+    generator->generate_const_scalar_op("abs", 'r', "r", "std::abs(a)");
+    generator->generate_const_scalar_op("abs", 'i', "i", "std::abs(a)");
+
+    // Complex number manipulation functions
+    generator->generate_const_scalar_op("complex", 'c', "rr", "primitives::Complex(a, b)");
+    generator->generate_const_scalar_op("polar", 'c', "rr", "std::polar<double>(a, b)");
+    generator->generate_const_scalar_op("real", 'r', "c", "std::real<double>(a)");
+    generator->generate_const_scalar_op("imag", 'r', "c", "std::imag<double>(a)");
+    generator->generate_const_scalar_op("arg", 'r', "c", "std::arg<double>(a)");
+    generator->generate_const_scalar_op("norm", 'r', "c", "std::norm<double>(a)");
+    generator->generate_const_scalar_op("conj", 'c', "c", "std::conj<double>(a)");
+}
+
+
+void generate_v1x_specific_functions(func_gen::Generator* generator) {
+    // Basic scalar arithmetic operators
+    generator->generate_const_scalar_op("operator+", 's', "ss", "a + b");
+    generator->generate_const_scalar_op("operator//", 'i', "ii", "div_floor(a, b)");
+
+    // Bitwise operators
+    generator->generate_const_scalar_op("operator>>>", 'i', "ii", "(int64_t)(((uint64_t)a) >> b)");
+}
+
+
+void test_v1x(const std::string &header_file_path, const std::string &source_file_path) {
+    std::unique_ptr<func_gen::Generator> generator{};
```

Comment:
you don't have to use a pointer I believe?
create the generator on the stack, and pass a reference in generate_common_functions and generate_v1x_specific_functions instead

Reply:
Yes, good point, thanks! Done.

I tend to think that I always have to use pointers and dynamic memory to work with runtime polymorphism.